### PR TITLE
ocamlPackages: Upgrade ocamlformat and ppxlib

### PR DESCRIPTION
**Pull request by Julow**
###### Motivation for this change

Add the newer versions of OCamlformat. Ppxlib is a dependency.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/ocaml-modules/ppxlib/default.nix [hunk 1]
```diff
@@ -1,21 +1,30 @@
 { lib, fetchFromGitHub, buildDunePackage, ocaml
-, version ? if lib.versionAtLeast ocaml.version "4.07" then "0.15.0" else "0.13.0"
+, version ? if lib.versionAtLeast ocaml.version "4.07" then "0.18.0" else "0.13.0"
```
Comment (sternenseemann) [inline pkgs/development/ocaml-modules/ppxlib/default.nix:2]: leave the default version at `0.15.0`, `>= 0.16.0` breaks a lot of packages currently.

File: pkgs/development/tools/ocaml/ocamlformat/generic.nix [hunk 2]
```diff
@@ -39,7 +41,23 @@ buildDunePackage rec {
   useDune2 = true;
 
   buildInputs =
-    if lib.versionAtLeast version "0.14"
+    if lib.versionAtLeast version "0.15.1"
```
Comment (sternenseemann) [inline pkgs/development/tools/ocaml/ocamlformat/generic.nix:44]: Any special reason for having 0.15.0 and 0.15.1?

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/development/ocaml-modules/ppxlib/default.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/ocaml-modules/ppxlib/default.nix:2]: To be precise this currently breaks:

```
dune-release jackline libbap patdiff python37Packages.bap python38Packages.bap python39Packages.bap satysfi
```

Comment (Julow) [reply pkgs/development/ocaml-modules/ppxlib/default.nix]: Fixed, thanks

File: pkgs/development/tools/ocaml/ocamlformat/generic.nix (hunk 2 above)
Comment (Julow) [reply pkgs/development/tools/ocaml/ocamlformat/generic.nix:44]: The only difference between 0.15.0 and 0.15.1 is the change in dependency (ocaml-migrate-parsetree vs ppxlib) but the exact version should be specified and checked into projects that use this tool. This means that every versions of OCamlformat are required.

Comment (sternenseemann) [reply pkgs/development/tools/ocaml/ocamlformat/generic.nix:44]: Alright fair, bit annoying that this means we specify a few not strictly required buildInputs for 0.15.1, but doesn't matter too much I think.

**Comment by sternenseemann** [comment]
@GrahamcOfBorg build jackline dune-release libbap satysfi patdiff

**Comment by sternenseemann** [comment]
Result of `nixpkgs-review pr 109288` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>17 packages built:</summary>
  <ul>
    <li>acgtk</li>
    <li>dune-release</li>
    <li>eff</li>
    <li>flow</li>
    <li>framac</li>
    <li>jackline</li>
    <li>libbap</li>
    <li>ocaml-crunch</li>
    <li>ocamlformat (ocamlformat_0_16_0)</li>
    <li>ocamlformat_0_15_1</li>
    <li>patdiff</li>
    <li>python37Packages.bap</li>
    <li>python38Packages.bap</li>
    <li>python39Packages.bap</li>
    <li>satysfi</li>
    <li>stog</li>
    <li>why3</li>
  </ul>
</details>

**Review by sternenseemann — APPROVED** [review]
I think this is good, you might want to squash e2ccb362a04762c20dd1cfa181ba3441cf1e6767 and 70558af832952fdc50945985e5412d58a9c1c0f9 into one commit.

**Comment by Julow** [comment]
Just squashed.

**Comment by vbgl** [comment]
Merged into master as bf010dac668c26989a6ffef967730076eb46ed12.

I’ve made two changes to the first commit. I’ve first increased the `min_version` bound that was too optimistic. Then I’ve ensured that the versions of `ppxlib` that were already there were not modified by this commit (in particular, the order of the `buildInputs` is preserved): this trivially ensures that the thousands of packages that use `ppxlib` are not affected by this unrelated addition.

**Comment by Julow** [comment]
Thanks !

**Pull request closed without merging**